## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
# Generate a random array
arr = np.random.randint(1,15, size=12)
arr

array([ 5,  9,  9, 12, 10,  8,  3,  5,  9,  9,  3,  4], dtype=int32)

In [3]:
# Option 1

# Filter data and multiply by -1
arr[(arr >= 3) & (arr <= 8)] *= -1
arr

array([-5,  9,  9, 12, 10, -8, -3, -5,  9,  9, -3, -4], dtype=int32)

In [4]:
# Option 2

# Define a mask (negate only where mask is True)
mask = np.where((arr >= 3) & (arr <= 8), True, False)

# Compute negative values where mask is True
res = np.negative(arr, where=mask, out=arr)
res

array([-5,  9,  9, 12, 10, -8, -3, -5,  9,  9, -3, -4], dtype=int32)

**2.** Заменить максимальный элемент случайного массива на 0

In [5]:
# Generate a random array
rand_arr = np.random.randint(-5,20, size=15)
rand_arr

array([ 7, -1, -4, 14, -3,  0,  5,  8,  2, 19, 19, 18,  6,  2, 16],
      dtype=int32)

In [6]:
# Find an index of maximum
rand_arr[rand_arr.argmax()] = 0
rand_arr

array([ 7, -1, -4, 14, -3,  0,  5,  8,  2,  0, 19, 18,  6,  2, 16],
      dtype=int32)

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [7]:
# Generate random arrays
mass1 = np.random.randint(1, 10, (3,5))
mass2 = np.random.randint(1, 12, (5,3))
print(mass1)
print(mass2)

[[4 9 6 5 3]
 [6 3 7 1 4]
 [8 7 8 8 1]]
[[ 8 11 10]
 [ 4  1  3]
 [ 5 10  5]
 [ 3  3  2]
 [ 6  4  1]]


In [8]:
np.matmul(mass1, mass2)

array([[131, 140, 110],
       [122, 158, 110],
       [162, 203, 158]], dtype=int32)

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [9]:
# Generate random arrays
massA = np.random.randint(0, 10, (8,3))
massB = np.random.randint(0, 10, (2,2))
print(massA)
print(massB)

[[0 4 6]
 [1 4 0]
 [7 3 5]
 [8 6 0]
 [2 9 0]
 [3 0 6]
 [9 8 6]
 [9 4 8]]
[[0 7]
 [1 8]]


In [10]:
def strB_in_strA (massA, massB):
    for RowB in massB:
        for RowA in massA:
            if all(item in RowA for item in RowB):
                print(f'{RowB} in {RowA}')

strB_in_strA(massA, massB)

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [11]:
# Generate random arrays
mass = np.random.randint(0, 4, (10,3))

# Compares if adjacent columns have the same values, returning True only if all elements in a row are equal
mass_compared = np.logical_and.reduce(mass[:,1:] == mass[:,:-1], axis=1)

# Creates a new array where not all elements in the row are identical
mass_new = mass[~mass_compared]

print(mass)
print(mass_new)

[[1 2 0]
 [2 1 3]
 [0 3 2]
 [3 2 1]
 [1 2 0]
 [3 3 3]
 [1 2 3]
 [1 3 3]
 [1 0 0]
 [0 3 1]]
[[1 2 0]
 [2 1 3]
 [0 3 2]
 [3 2 1]
 [1 2 0]
 [1 2 3]
 [1 3 3]
 [1 0 0]
 [0 3 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [12]:
# Generate random arrays
mass = np.random.randint(0, 4, (8, 2))

# Convert each row in array into a tuple
mass_tuple = [tuple(row) for row in mass]

# Remove duplicate rows from the array
mass_res = np.unique(mass_tuple,axis=0)

print(mass)
print(mass_res)

[[3 3]
 [0 1]
 [1 3]
 [0 2]
 [3 0]
 [0 1]
 [1 2]
 [0 2]]
[[0 1]
 [0 2]
 [1 2]
 [1 3]
 [3 0]
 [3 3]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [13]:
# Generate random arrays
arr = np.random.randint(0, 5, (4, 3))
arr

array([[4, 4, 4],
       [1, 1, 3],
       [4, 2, 4],
       [1, 4, 3]], dtype=int32)

In [14]:
# Option 1

def product_nonzero_diag(matrix_rectangle):
    product = 1
    for i in range(min(len(arr[0]), len(arr))):
        if arr[i][i] != 0:
            product *= arr[i][i]
    print(f'Product of nonzero elements on the diagonal: {product}')

product_nonzero_diag(arr)

Product of nonzero elements on the diagonal: 16


In [15]:
# Option 2 (numpy)

product = np.prod(np.diag(arr)[np.diag(arr)!=0])
print(f'Product of nonzero elements on the diagonal: {product}')

Product of nonzero elements on the diagonal: 16


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [16]:
# Generate random arrays
arr1 = np.random.randint(0, 5, 4)
arr2 = np.random.randint(0, 5, 4)
print(arr1)
print(arr2)

[0 1 3 2]
[4 1 2 2]


In [17]:
# Option 1 

sorted(arr1) == sorted(arr2)

False

In [18]:
# Option 2 (numpy)

print(np.array_equal(np.unique(arr1), np.unique(arr2)))

False


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [21]:
# Generate random arrays
arr = np.random.randint(0, 8, 10)
arr

array([1, 0, 1, 4, 7, 7, 7, 3, 7, 1], dtype=int32)

In [22]:
# Option 1

list_elem = [arr[i+1] for i in range(len(arr)) if arr[i] == 0]
print(max(list_elem))

1


In [23]:
# Option 2 (numpy)

print(np.max(arr[1:][(arr==0)[:-1]]))

1


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [24]:
#Generate random arrays
x = np.random.randint(0, 6, 8)
x

array([0, 1, 4, 3, 3, 1, 3, 0], dtype=int32)

In [25]:
# Option 1

def count_repeats_with_lists(x):
    unique_numbers = []
    counts = []

    for num in x:
        if num in unique_numbers:
            index = unique_numbers.index(num)
            counts[index] += 1
        else:
            unique_numbers.append(num)
            counts.append(1)
            
    return (np.array(unique_numbers), np.array(counts))

print(count_repeats_with_lists(x))

(array([0, 1, 4, 3], dtype=int32), array([2, 2, 1, 3]))


In [26]:
# Option 2 (numpy)

unic_num = tuple(np.unique(x, return_counts = True))
unic_num

(array([0, 1, 3, 4], dtype=int32), array([2, 2, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [27]:
# Generate random arrays
X = np.random.randint(1, 6, size=(3,4))
Y = np.random.randint(0, 9, size=(3,4))
print(X)
print(Y)

[[1 3 3 1]
 [2 3 3 1]
 [4 1 3 3]]
[[8 1 4 8]
 [6 4 4 4]
 [4 3 5 0]]


In [28]:
# Option 1 (numpy)

# Calculate the Euclidean distance matrix
dist_matrix = np.sqrt(((X[:, np.newaxis] - Y[np.newaxis, :]) ** 2).sum(axis=2))

print(dist_matrix)

[[10.14889157  6.          3.74165739]
 [ 9.48683298  5.19615242  3.        ]
 [ 6.4807407   3.87298335  4.12310563]]


In [29]:
#Option 2 (scipy)

from scipy.spatial.distance import cdist
#Calculate the Euclidean distance matrix
dist_matrix = cdist(X, Y, metric='euclidean')

print(dist_matrix)

[[10.14889157  6.          3.74165739]
 [ 9.48683298  5.19615242  3.        ]
 [ 6.4807407   3.87298335  4.12310563]]


In [30]:
# Option 1 (numpy)
import timeit
code_to_test = """

import numpy as np
X = np.random.randint(1, 6, size=(3,4))
Y = np.random.randint(0, 9, size=(3,4))
dist_matrix = np.sqrt(((X[:, np.newaxis] - Y[np.newaxis, :]) ** 2).sum(axis=2))
"""
execution_time = timeit.timeit(code_to_test, number=1000)
print(f"Average time of execution numpy: {execution_time:.4f} sec")

Average time of execution numpy: 0.0142 sec


In [31]:
#Option 2 (scipy)

import timeit
code_to_test = """

import numpy as np
X = np.random.randint(1, 6, size=(3,4))
Y = np.random.randint(0, 9, size=(3,4))
from scipy.spatial.distance import cdist
dist_matrix = cdist(X, Y, metric='euclidean')
"""
execution_time = timeit.timeit(code_to_test, number=1000)
print(f"Average time of execution scipy: {execution_time:.4f} sec")

Average time of execution scipy: 0.0143 sec


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [32]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [33]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [34]:
# your code
average_calories = np.mean(calorie_stats).round(0)
print(average_calories)

107.0


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [35]:
# your code
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [36]:
# your code
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [37]:
for i in reversed(range(1,20)):
    print(f'{i} percentile - {np.percentile(calorie_stats, i):.1f} calories')

19 percentile - 100.0 calories
18 percentile - 100.0 calories
17 percentile - 99.2 calories
16 percentile - 91.6 calories
15 percentile - 90.0 calories
14 percentile - 90.0 calories
13 percentile - 90.0 calories
12 percentile - 90.0 calories
11 percentile - 90.0 calories
10 percentile - 90.0 calories
9 percentile - 90.0 calories
8 percentile - 90.0 calories
7 percentile - 83.2 calories
6 percentile - 75.6 calories
5 percentile - 70.0 calories
4 percentile - 70.0 calories
3 percentile - 55.6 calories
2 percentile - 50.0 calories
1 percentile - 50.0 calories


In [38]:
caloriesCM = 60
for i in reversed(range(1,11)):
    if np.percentile(calorie_stats, i) < caloriesCM:
            nth_percentile = np.percentile(calorie_stats, i+1)
            print(f'{i+1} percentile - {nth_percentile:.1f} calories')
            break

4 percentile - 70.0 calories


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [39]:
# your code
caloriesCM = 60
more_calories = np.count_nonzero(calorie_stats > caloriesCM) / len(calorie_stats)
print(f'Percent of CrunchieMunchies with more than {caloriesCM} calories - {more_calories:.1%}')

Percent of CrunchieMunchies with more than 60 calories - 96.1%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [40]:
# Standard deviation
calorie_std = np.std(calorie_stats)
print(f'Standard deviation - {calorie_std:.0f} calories')
print(f'Scatter in data is medium - {(calorie_std / average_calories):.1%}')

Standard deviation - 19 calories
Scatter in data is medium - 18.1%


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

**Competitors' CrunchieMunchies contain 107 the average calories, 110 - the median. The standard deviation for the data indicate medium variability.
MyCrunch's product contain much less calories than the average: 96% of competitors' CrunchieMunchies contain more than 60 calories.
This can be used in marketing to highlight the product's low-calorie and light nature. That is especially important for consumers who tend a healthy lifestyle.**